### ISSUE TITLE
Support [[]]-style Links in Markdown

### ISSUE THREAD

**Issue by iamrecursion** [prettier/prettier#9161]
**Prettier 2.1.1**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEB1AFggBATwgVy3QEMA3OLAZwgFsKAbaAcywAcAnCSimEmLSLGIBLKJSzEs7OADM40qGDgBaeqIDWVGDnoMNWANoG16numGUAupYA0baTBjD5WAO7D69LACNpxTbx6UJoQ5OxYNPj0Tqy6WGpQcJQAdFgAKubi8FAAJlkQPn4BmNxY6qJ5WBAy8RriMhDhxFA4WDAQEF68xPzSaklY+KzQbZg0yQA6UFMAklAS88I5cMRero30OXaBpZToBJs+PH7wObWmWs05xOxnBug4rPIJ6lYAFOiOrJRIAPS-cAAHsQaLE4MlIDQAJQSaR2K5udjEVhPM7EcTQCgIJzaZIgGwgCCsJzQSjIUA3TiuAAKNwQZJQq1cxBwZIJvmIYFMMAAyqxOaImMgYOx8HACXAaN44DlljkADLNJj4YhMOAAMUaNB6TigQsZ+Ha+JAnxo9AwwnglH5Sh59MtwlIlpwyHAlDZIFE3HYMGpSKY2uQMlW3AJACtKICAEJIrlwXkguDy0RwIMh8UgCOAnmC3QARXwEHgafooZA-PY3td2vY6hyEFcUGNHFEMFQS14yAAHAAGAkcLhwVBI1iupksj0AR0L8D9RIZIHRykSMplxuk0+E0j9qsDSGDpYz3BowmFoqPubgBaLqf36YJMGI3nbOU7SAATA+kR5BQBhWh7iASQAKzGvg3BpE+DIHmWpBinMyywDyYDsMIxIAIK5Dy2i6CW3AAL74UAA)
```sh
--parser markdown
--prose-wrap always
```

**Input:**
```markdown
When you have some long prose that contains a reference-link style link [[like this]], prettier will break the link over multiple lines. This tends to break these kinds of links for any tool that relies upon them.

In an ideal world, these should be treated like standard [hyperlinks](https://example.com) are, and wrapped as one entity.
```

**Output:**
```markdown
When you have some long prose that contains a reference-link style link [[like
this]], prettier will break the link over multiple lines. This tends to break
these kinds of links for any tool that relies upon them.

In an ideal world, these should be treated like standard
[hyperlinks](https://example.com) are, and wrapped as one entity.

```

**Expected behavior:**
The `[[]]`-style links should be treated as a single block, much like standard markdown link syntax is today.

I do recognise that `[[]]`-style links aren't commonmark-compliant, but they're a common format for internal links in extended wiki-style markdown used by tools like [obsidian](https://obsidian.md/), [foam](https://github.com/foambubble/foam), [azure devops wiki](https://docs.microsoft.com/en-us/azure/devops/project/wiki/wiki-markdown-guidance?view=azure-devops) and so on. Without it, it's common that formatting markdown using this syntax will break links at random.

**Comment by iamrecursion** [comment]
If I were to look into implementing this myself, would it be something the project is interested in supporting? If so, where's a good starting point to look?

**Comment by evilebottnawi** [comment]
I think we need remark plugin/option for this https://github.com/remarkjs/remark, because we need parse (AST), maybe it is worth opening a problem about it in remark repo and ask developers about it

**Comment by iamrecursion** [comment]
I've opened remarkjs/remark#531 with the question.

**Comment by iamrecursion** [comment]
> In remark (and unified as a whole), we typically do everything with plugins...
[This](https://github.com/remarkjs/remark/issues/531) would imply that we should create a plugin for it! The existing plugin linked from that comment looks a touch more complex than would be needed for formatting purposes. As far as I understand it, all that prettier would need to know is "this is a wiki link, do not break".

What do you think, @evilebottnawi?

**Comment by evilebottnawi** [comment]
Yes, we can put this plugin inside code

**Comment by iamrecursion** [comment]
Cool! I'm not going to officially 'claim' this issue, but if I get some spare time soon I'll give it a shot!

**Comment by iamrecursion** [comment]
Alrighty I've been implementing this and I've run into a little snag. No matter what I do in `printer-markdown.js`, the entity ends up printed as a separate paragraph. Is there something obvious I'm missing? As far as I can see, I'm behaving like a link in that I'm returning the output string, but a standard markdown link remains in the same content block, and the wiki link doesn't.

```js
    case "wikiLink": {
      return concat(["[[", node.value, "]]"]);
    }
```

Any thoughts @evilebottnawi?

**Comment by iamrecursion** [comment]
Ah, please ignore the above. I'd forgotten to add it to `INLINE_NODE_TYPES`.

**Comment by iamrecursion** [comment]
PR open at #9275.

**Comment by iamrecursion** [comment]
Closed in #9275!
